### PR TITLE
Add `output_dtype` and `output_ndim` arguments to Pipeline constructor

### PR DESCRIPTION
**Pull request by szalpal**
## Category:

**New feature** (*non-breaking change which adds functionality*)

## Description:
In order to push forward the autoconfig feature in the DALI Backend, we need `output_ndim` and `output_dtype` arguments. The reason is that DALI Backend needs a way to determine these two parameters. A better solution would be a type and dimensionality inference in DALI, but this is a big feature.

`output_ndim` defines, how many dimensions each output might have. `output_dtype` define, what is its type. For more information on how these two arguments work, please refer to their documentation in `pipeline.py` file.

This PR adds the two arguments and all the code necessary to pipe them through Python layer into the Pipeline class. I'm modifying the output descriptor, which previously was a `pair<string, string>`, denoting name of the output and the device of the output. After the change it is a `PipelineOutputDesc` with all the data. It's named `Pipeline...`, because Executor also has its own output descriptor, but it is limited to the output name.

The two parameters needed to be included in the proto definition, since we want them to be accessible after serialization and deserialization.

Lastly, I've added `output_dtype` and `output_ndim` to the AudioDecoder test, so that it has bigger coverage and more real-life scenario than the unit tests in `test_pipeline.py`.

### Affected modules and functionalities:
`pipeline.py`, python bindings, protobuf definition, Pipeline & Workspace classes, tests

### Key points relevant for the review:
1. Documentation (is it sufficient?)
2. Protobuf record for the `dtype`. Could it be applied in a more elegant way?
3. Tests in `pipeline.py`. Did I covered all necessary cases? Shall any other test case be added?

### Tests
- [ ] Existing tests apply
- [ ] New tests added
  - [x] Python tests
  - [ ] GTests
  - [ ] Benchmark
  - [ ] Other
- [ ] N/A

### Documentation
- [ ] Existing documentation applies
- [x] Documentation updated
  - [ ] Docstring
  - [ ] Doxygen
  - [ ] RST
  - [ ] Jupyter
  - [ ] Other
- [ ] N/A

#### Requirements
- [ ] Implements new requirements
- [ ] Affects existing requirements
- [ ] N/A

**REQ IDs**: N/A

**JIRA TASK**: N/A

### CODE REVIEW

File: dali/pipeline/pipeline.cc [hunk 1]
```diff
@@ -554,32 +577,53 @@ void Pipeline::SetCompletionCallback(ExecutorBase::ExecutorCallback cb) {
...
-    }
+  DALI_ENFORCE(built_, "\"Build()\" must be called prior to executing the pipeline.");
+  try {
+    executor_->Outputs(ws);
+  } catch (std::exception &e) {
+    throw std::runtime_error(make_string("Critical error in pipeline:\n", std::string(e.what()),
+                                         "\nCurrent pipeline object is no longer valid."));
+  } catch (...) {
+    throw std::runtime_error("Unknown critical error in pipeline.");
+  }
```
Comment (szalpal) [inline dali/pipeline/pipeline.cc:610]: Indentation fix

File: dali/pipeline/pipeline.cc [hunk 2]
```diff
@@ -554,32 +577,53 @@ void Pipeline::SetCompletionCallback(ExecutorBase::ExecutorCallback cb) {
...
-    }
+  DALI_ENFORCE(built_, "\"Build()\" must be called prior to executing the pipeline.");
+  try {
+    executor_->ShareOutputs(ws);
+  } catch (std::exception &e) {
+    throw std::runtime_error(make_string("Critical error in pipeline:\n", std::string(e.what()),
+                                         "\nCurrent pipeline object is no longer valid."));
+  } catch (...) {
+    throw std::runtime_error("Unknown critical error in pipeline.");
+  }
```
Comment (szalpal) [inline dali/pipeline/pipeline.cc:624]: Indentation fix

**Comment by szalpal** [comment]
!build

File: dali/pipeline/data/types.h [hunk 3]
```diff
@@ -214,6 +214,68 @@ inline const char *GetBuiltinTypeName(DALIDataType t) {
   }
 }
 
+inline DALIDataType GetBuiltinDataType(const std::string &type_name) {
```
Comment (klecki) [inline dali/pipeline/data/types.h:217]: As mentioned, I find the conversion in this direction weird, as this string representation is intended mostly for visual presentation (docs, erros) rather than serialization.

I would suggest sticking with DALIDataType directly.

File: dali/pipeline/pipeline.h [hunk 4]
```diff
@@ -498,6 +515,12 @@ class DLL_PUBLIC Pipeline {
   int GetNextLogicalId();
   int GetNextInternalLogicalId();
 
+  /**
+   * Validate, that the outputs from the Pipeline match the criteria.
```
Comment (klecki) [inline dali/pipeline/pipeline.h:519]: What criteria? :)

File: dali/pipeline/pipeline.cc [hunk 5]
```diff
@@ -94,10 +94,12 @@ void DeserializeOpSpec(const dali_proto::OpDef &def, OpSpec *spec) {
 Pipeline::Pipeline(const string &serialized_pipe, int batch_size, int num_threads, int device_id,
                    bool pipelined_execution, int prefetch_queue_depth, bool async_execution,
                    size_t bytes_per_sample_hint, bool set_affinity, int max_num_stream,
-                   int default_cuda_stream_priority, int64_t seed)
-        : built_(false), separated_execution_(false) {
-    dali_proto::PipelineDef def;
-    DALI_ENFORCE(DeserializePipeline(serialized_pipe, def), "Error parsing serialized pipeline.");
+                   int default_cuda_stream_priority, int64_t seed,
+                   const std::vector<DALIDataType> &output_dtype,
+                   const std::vector<int> &output_ndim)
```
Comment (klecki) [inline dali/pipeline/pipeline.cc:99]: Looks like those are not used here.

File: dali/pipeline/pipeline.h [hunk 6]
```diff
@@ -251,18 +255,21 @@ class DLL_PUBLIC Pipeline {
...
   DLL_PUBLIC OpNode * GetOperatorNode(const std::string& name);
 
+  ///@{
   /**
    * @brief Performs some checks on the user-constructed pipeline, setups data
    * for intermediate results, and marks as ready for execution. The input
    * vector specifies the name and device of the desired outputs of the pipeline.
    */
-  DLL_PUBLIC void Build(vector<std::pair<string, string>> output_names);
+  DLL_PUBLIC void Build(const std::vector<std::pair<string, string>>& output_names);
```
Comment (klecki) [inline dali/pipeline/pipeline.h:264]: Do we need to keep the old overload?

File: dali/pipeline/proto/dali.proto [hunk 7]
```diff
@@ -35,6 +35,8 @@ message InputOutput {
   required string device = 2;
   required bool is_argument_input = 3 [default = false];
   optional string arg_name = 4;
+  optional string dtype = 5;
```
Comment (klecki) [inline dali/pipeline/proto/dali.proto:38]: How about using int here?

File: dali/pipeline/pipeline.h [hunk 8]
```diff
@@ -101,7 +104,8 @@ class DLL_PUBLIC Pipeline {
                       int prefetch_queue_depth = 2, bool async_execution = true,
                       size_t bytes_per_sample_hint = 0, bool set_affinity = false,
                       int max_num_stream = -1, int default_cuda_stream_priority = 0,
-                      int64_t seed = -1);
+                      int64_t seed = -1, const std::vector<DALIDataType> &output_dtype = {},
```
[the remark this answers is not in the document]
Comment (klecki) [reply dali/pipeline/pipeline.h:107]: Looks like we are providing them via Build and this is just leftover. But I'm leaving the comment above, as what I said still stands for other features and future reference.

File: dali/python/nvidia/dali/pipeline.py [hunk 9]
```diff
@@ -152,14 +152,32 @@ def create_pipeline():
     by decorating them with `@dali.pickling.pickle_by_value`. It may be especially useful when
     working with Jupyter notebook to work around the issue of worker process being unable to import
     the callback defined as a global function inside the notebook.
+`output_dtype` : nvidia.dali.types or list of those, default = None
+    With this argument, you may declare, what data type you expect in the given output. You shall
+    pass a list of values from module:`nvidia.dali.types`, each element in the list corresponding to
```
Comment (klecki) [inline dali/python/nvidia/dali/pipeline.py:157]:
```suggestion
    pass a list of types.DALIDataType, each element in the list corresponding to
```
And please link to that type. Also adjust it in all the occurrences here in the doc and in the error messages. It can't be types.SampleInfo right?

File: dali/python/nvidia/dali/pipeline.py [hunk 10]
```diff
@@ -1203,6 +1244,22 @@ def iter_setup(self):
         data from NumPy arrays."""
         pass
 
+    def _generate_build_args(self, output_dtype, output_ndim):
+        ret = []
+        num_outputs = len(self._names_and_devices)
```
Comment (klecki) [inline dali/python/nvidia/dali/pipeline.py:1249]: Just curious why do you pass output_dtype and output_ndim but not _names_and_devices as arguments? (small nitpick).

File: dali/python/nvidia/dali/pipeline.py [hunk 11]
```diff
@@ -1203,6 +1244,22 @@ def iter_setup(self):
...
+        output_dtype = [output_dtype] * num_outputs if type(
+            output_dtype) is not list else output_dtype
+        output_ndim = [output_ndim] * num_outputs if type(output_ndim) is not list else output_ndim
+        if not (len(output_dtype) == len(output_ndim) == num_outputs):
+            raise RuntimeError(
+                f"Inconsistent output description. Length of provided output descriptions do not match. \n"
+                f"Expected num_outputs={num_outputs}.\nReceived:\noutput_dtype={output_dtype}\noutput_ndim={output_ndim}")
+
+        for nd, dtype, ndim in zip(self._names_and_devices, output_dtype, output_ndim):
```
Comment (klecki) [inline dali/python/nvidia/dali/pipeline.py:1258]: What is names_and_devices? Is it tuple? Would this work:
```suggestion
        for (name, device), dtype, ndim in zip(self._names_and_devices, output_dtype, output_ndim):
```

File: dali/test/python/test_operator_decoders_audio.py [hunk 12]
```diff
@@ -50,16 +50,19 @@ def create_test_files():
...
 
+
 class DecoderPipeline(Pipeline):
   def __init__(self):
-    super(DecoderPipeline, self).__init__(batch_size=8, num_threads=3, device_id=0,
-                                          exec_async=True, exec_pipelined=True)
+    super().__init__(batch_size=8, num_threads=3, device_id=0, exec_async=True, exec_pipelined=True,
+                     output_dtype=[types.INT16, types.INT16, types.INT16, types.FLOAT,
+                                   types.FLOAT, types.FLOAT, types.FLOAT, types.FLOAT],
+                     output_ndim=[2, 2, 1, 1, 0, 0, 0, 0])
```
Comment (klecki) [inline dali/test/python/test_operator_decoders_audio.py:59]: Just a side-note, that this being separate arguments (and separate from output specification) isn't the nicest looking thing we did.

File: dali/util/pybind.h [hunk 13]
```diff
@@ -231,5 +231,36 @@ static DLMTensorPtr DLMTensorPtrFromCapsule(py::capsule &capsule) {
   return {DLMTensorRawPtrFromCapsule(capsule), DLMTensorPtrDeleter};
 }
 
+
+/**
```
Comment (klecki) [inline dali/util/pybind.h:235]: I guess if you export the PipelineOutputDesc this won't be necessary. I am not sure if it's worth to add the code with more comments rather than just using get<i> by hand either way.

**Review by klecki — COMMENTED** [review]
Posting some comments.
I think there should be one point of introducing the PipelineOutputDesc, we probably can get rid of other APIs.

It looks like we do not propagate the information when serializing, unless I misread something. Please see the serialize in pipeline.py.

File: dali/pipeline/pipeline.cc [hunk 14]
```diff
@@ -534,7 +544,16 @@ void Pipeline::Build(vector<std::pair<string, string>> output_names) {
 }
 
 void Pipeline::SetOutputNames(const vector<std::pair<string, string>> &output_names) {
-  output_names_ = output_names;
+  DALI_ENFORCE(output_descs_.empty() || output_descs_.size() == output_names.size());
```
[the remark this answers is not in the document]
Comment (klecki) [reply dali/pipeline/pipeline.cc:547]: I guess one solution would be to use PipelineOutputDesc here and remove the pair overload.

File: dali/pipeline/data/types.h (hunk 3 above)
Comment (szalpal) [reply dali/pipeline/data/types.h]: Done

File: dali/python/nvidia/dali/pipeline.py (hunk 11 above)
Comment (szalpal) [reply dali/python/nvidia/dali/pipeline.py]: Yes it would

File: dali/pipeline/pipeline.h (hunk 4 above)
Comment (szalpal) [reply dali/pipeline/pipeline.h:532]: The criteria ;)

File: dali/python/backend_impl.cc [hunk 15]
```diff
@@ -1569,13 +1569,23 @@ PYBIND11_MODULE(backend_impl, m) {
...
+         })
     .def("Build",
-        [](Pipeline *p) {
-          p->Build();
-          })
+         [](Pipeline *p, const std::vector<
+                           std::tuple<std::string  /* name */,
+                                      std::string  /* device */,
+                                      DALIDataType /* dtype */,
+                                      int          /* ndim */>
```
[the remark this answers is not in the document]
Comment (szalpal) [reply dali/python/backend_impl.cc]: That's good point. I'll have a go with `@dataclass`, we'll see how it works

File: dali/python/nvidia/dali/pipeline.py [hunk 16]
```diff
@@ -218,6 +236,29 @@ def __init__(self, batch_size = -1, num_threads = -1, device_id = -1, seed = -1,
         else:
             raise TypeError("Expected prefetch_queue_depth to be either int or Dict[int, int]")
 
+        # Assign and validate output_dtype
+        if type(output_dtype) is list:
+            for dtype in output_dtype:
+                assert type(dtype) is types.DALIDataType or dtype is None, \
```
[the remark this answers is not in the document]
Comment (szalpal) [reply dali/python/nvidia/dali/pipeline.py]: Done

File: dali/util/pybind.h (hunk 13 above)
Comment (szalpal) [reply dali/util/pybind.h:235]: I'll have a go with `@dataclass`, we'll see if this will stay or not

File: dali/pipeline/pipeline.cc (hunk 14 above)
[the remark this answers is not in the document]
Comment (szalpal) [reply dali/pipeline/pipeline.cc]: I'd leave the pair overload. Removing it would cause many changes in the test code. I've renamed the function to `SetOutputDescs`.

File: dali/python/nvidia/dali/pipeline.py (hunk 10 above)
Comment (szalpal) [reply dali/python/nvidia/dali/pipeline.py:1262]: Good point

File: dali/python/nvidia/dali/pipeline.py (hunk 9 above)
Comment (szalpal) [reply dali/python/nvidia/dali/pipeline.py:157]: Done

**Review by klecki — CHANGES_REQUESTED** [review]
As already mentioned, the `output_dtype` and `output_ndim` are not used in Pipeline constructor, and as I suggested, we should not be stretching the Pipeline constructor with more arguments.

File: dali/pipeline/pipeline.h [hunk 17]
```diff
@@ -101,7 +104,8 @@ class DLL_PUBLIC Pipeline {
                       int prefetch_queue_depth = 2, bool async_execution = true,
                       size_t bytes_per_sample_hint = 0, bool set_affinity = false,
                       int max_num_stream = -1, int default_cuda_stream_priority = 0,
-                      int64_t seed = -1);
+                      const std::vector<DALIDataType> &output_dtype = {},
+                      const std::vector<int> &output_ndim = {}, int64_t seed = -1);
```
[the remark this answers is not in the document]
Comment (szalpal) [reply dali/pipeline/pipeline.h]: Done

File: dali/python/nvidia/dali/pipeline.py [hunk 18]
```diff
@@ -218,6 +236,33 @@ def __init__(self, batch_size = -1, num_threads = -1, device_id = -1, seed = -1,
         else:
             raise TypeError("Expected prefetch_queue_depth to be either int or Dict[int, int]")
 
+        # Assign and validate output_dtype
+        if type(output_dtype) is list:
+            for dtype in output_dtype:
+                if type(dtype) is not types.DALIDataType and dtype is not None:
+                    raise TypeError(
+                        f"`output_dtype` must be either: a type from nvidia.dali.types.DALIDataType, a list of these or None. Found type {type(dtype)} in the list.")
```
Comment (klecki) [inline dali/python/nvidia/dali/pipeline.py:244]: Hmm, technically those are enum values :)
```suggestion
                        f"`output_dtype` must be either: None, a value of nvidia.dali.types.DALIDataType type or a list of these. Found type {type(dtype)} in the list.")
```
I know what you mean, but maybe something like this, or similar? (I also changed the order a bit, maybe it matches the doc a bit better).

Same below.

File: dali/python/backend_impl.cc [hunk 19]
```diff
@@ -1569,13 +1577,18 @@ PYBIND11_MODULE(backend_impl, m) {
...
-        [](Pipeline *p) {
-          p->Build();
-          })
+         [](Pipeline *p, const std::vector<OutputDesc> &outputs) {
+             std::vector<PipelineOutputDesc> build_args;
+             for (auto& out : outputs) {
+               build_args.emplace_back(to_struct<PipelineOutputDesc>(out));
+             }
+             p->Build(build_args);
+         })
```
Comment (klecki) [inline dali/python/backend_impl.cc:1590]: I know you've kept the overload in C++ for the purpose of not reworking the C++ API tests, but maybe we should keep only one option in Python. This is fully internal and used in one place, maybe remove the old overload?

File: dali/test/python/test_pipeline.py [hunk 20]
```diff
@@ -1654,3 +1654,102 @@ def test_invoke_serialize_error_handling_string():
...
+    both_correct = create_test_package(output_dtype=[types.UINT8], output_ndim=[3])
+    ndim_correct_dtype_wildcard = create_test_package(output_dtype=[None], output_ndim=[3])
+    dtype_correct_ndim_wildcard = create_test_package(output_dtype=types.UINT8)
+    dtype_incorrect = create_test_package(output_dtype=[types.FLOAT], output_ndim=[3])
+    ndim_incorrect = create_test_package(output_dtype=types.UINT8, output_ndim=0)
+    both_correct_one_list = create_test_package(output_dtype=types.UINT8, output_ndim=[3])
+    too_many_dtypes = create_test_package(output_dtype=[types.UINT8, types.FLOAT])
+    correct_dtypes_but_too_many = create_test_package(output_dtype=[types.UINT8, types.UINT8])
+    correct_ndims_but_too_many = create_test_package(output_ndim=[3, 3])
+    all_wildcards = create_test_package()
```
Comment (klecki) [inline dali/test/python/test_pipeline.py:1709]: How about passing incorrect values, like output_dtype=[int] or output_ndim=[-5.0]. etc. Also the specific defaults, that is -1 and DALI_NO_TYPE?

**Review by klecki — COMMENTED** [review]
Mostly nitpicks and two actual problems:
1. If we pass None as placeholder, should we return it back as `None` in the output_dtype or output_ndim getters in Python pipeline?
2. I think we may drop exposing one of the Build(...) in backend_impl if it is not (intended to be) used in Python.

File: dali/python/nvidia/dali/pipeline.py [hunk 21]
```diff
@@ -1203,6 +1257,22 @@ def iter_setup(self):
...
+        num_outputs = len(self._names_and_devices)
+        dtype = [self._output_dtype] * num_outputs if type(
+            self._output_dtype) is not list else self._output_dtype
+        ndim = [self._output_ndim] * num_outputs if type(self._output_ndim) is not list else self._output_ndim
+        if not (len(dtype) == len(ndim) == num_outputs):
+            raise RuntimeError(
+                f"Inconsistent output description. Length of provided output descriptions do not match. \n"
+                f"Expected num_outputs={num_outputs}.\nReceived:\noutput_dtype={dtype}\noutput_ndim={ndim}")
+
+        for (name, dev), dt, nd in zip(self._names_and_devices, dtype, ndim):
```
Comment (stiepan) [inline dali/python/nvidia/dali/pipeline.py:1271]: nitpick: You could go with list comprehension here instead of ret = [] and append to be more pythonic.

File: dali/test/python/test_pipeline.py [hunk 22]
```diff
@@ -1654,3 +1654,102 @@ def test_invoke_serialize_error_handling_string():
...
+    def ndim_dtype_matches(test_value, ref_value):
+        ref_value = ref_value if type(ref_value) is list else [ref_value]
+        for tv, ref in zip(test_value, ref_value):
+            if ref is None:
+                if tv != -1:
+                    return False
+            else:
+                if tv != ref:
+                    return False
+        return True
```
Comment (stiepan) [inline dali/test/python/test_pipeline.py:1669]: I guess this is good example to support @klecki proposal to get rid of -1.

File: dali/test/python/test_pipeline.py [hunk 23]
```diff
@@ -1654,3 +1654,102 @@ def test_invoke_serialize_error_handling_string():
 @raises(TypeError, "*define_graph*callable*")
 def test_invoke_serialize_error_handling_not_string():
     _identity_pipe().serialize(42)
+
+
+def check_dtype_ndim(dali_pipeline, output_dtype, output_ndim):
+    def ndim_dtype_matches(test_value, ref_value):
+        ref_value = ref_value if type(ref_value) is list else [ref_value]
```
Comment (stiepan) [inline dali/test/python/test_pipeline.py:1661]: If `type(ref_value) is list` is False you zip `[ref_value]` of len 1 with test_value effectively ignoring all the remaining outputs.

File: dali/pipeline/pipeline.h (hunk 4 above)
Comment (stiepan) [reply dali/pipeline/pipeline.h:529]: First rule of the validation criteria.

File: dali/pipeline/pipeline_output_desc.h [hunk 24]
```diff
@@ -0,0 +1,48 @@
...
+ * Note: the Executor also has an output descriptor inside. It is different than this one.
+ */
+struct PipelineOutputDesc {
+  std::string name, device;
+  DALIDataType dtype;
+  int ndim;
+
+  PipelineOutputDesc() = default;
+
+  PipelineOutputDesc(std::string name, std::string device, DALIDataType dtype,int ndim )
```
Comment (stiepan) [inline dali/pipeline/pipeline_output_desc.h:36]: nitpick: whitespaces

File: dali/pipeline/pipeline.cc [hunk 25]
```diff
@@ -421,11 +422,18 @@ void Pipeline::PropagateMemoryHint(OpNode &node) {
...
-void Pipeline::Build(vector<std::pair<string, string>> output_names) {
+void Pipeline::Build(const vector<std::pair<string, string>>& output_names) {
+  std::vector<PipelineOutputDesc> output_descs = {output_names.begin(), output_names.end()};
+  this->Build(output_descs);
+}
+
+void Pipeline::Build(std::vector<PipelineOutputDesc> output_descs) {
   DeviceGuard d(device_id_);
-  output_names_ = output_names;
+  output_descs_ = std::move(output_descs);
```
[the remark this answers is not in the document]
Comment (szalpal) [reply dali/pipeline/pipeline.cc:432]: Done

File: dali/python/nvidia/dali/pipeline.py [hunk 26]
```diff
@@ -1203,6 +1257,22 @@ def iter_setup(self):
         data from NumPy arrays."""
         pass
 
+    def _generate_build_args(self):
+        ret = []
+        num_outputs = len(self._names_and_devices)
+        dtype = [self._output_dtype] * num_outputs if type(
+            self._output_dtype) is not list else self._output_dtype
+        ndim = [self._output_ndim] * num_outputs if type(self._output_ndim) is not list else self._output_ndim
```
[the remark this answers is not in the document]
Comment (szalpal) [reply dali/python/nvidia/dali/pipeline.py:1265]: True that

File: dali/python/nvidia/dali/pipeline.py (hunk 21 above)
[the remark this answers is not in the document]
Comment (szalpal) [reply dali/python/nvidia/dali/pipeline.py:1271]: 👍🏻

File: dali/python/nvidia/dali/pipeline.py [hunk 27]
```diff
@@ -304,6 +349,14 @@ def gpu_queue_size(self):
         """The number of iterations processed ahead by the GPU stage."""
         return self._gpu_queue_size
 
+    def output_dtype(self) -> list:
+        """Data types expected at the outputs."""
+        return self._pipe.output_dtype()
+
+    def output_ndim(self) -> list:
+        """Number of dimensions expected at the outputs."""
+        return self._pipe.output_ndim()
```
[the remark this answers is not in the document]
Comment (szalpal) [reply dali/python/nvidia/dali/pipeline.py:358]: Yes

File: dali/python/nvidia/dali/pipeline.py (hunk 21 above)
Comment (szalpal) [reply dali/python/nvidia/dali/pipeline.py:1271]: Done

File: dali/pipeline/pipeline.cc [hunk 28]
```diff
@@ -554,32 +568,52 @@ void Pipeline::SetCompletionCallback(ExecutorBase::ExecutorCallback cb) {
   executor_->SetCompletionCallback(std::move(cb));
 }
 
+bool Pipeline::ValidateOutputs(const DeviceWorkspace &ws) const {
+  DALI_ENFORCE(
+      ws.NumOutput() == static_cast<int>(output_descs_.size()),
+      make_string("Inconsistent output description. Number of outputs does not match. Expected: ",
```
Comment (stiepan) [inline dali/pipeline/pipeline.cc:574]: Is this _Inconsistent output description_ though? I support that term in the error saying you specified different number of types than there are outputs, but here it was confusing to me. I'd drop the first sentence from the three error messages here.

Comment (szalpal) [reply dali/pipeline/pipeline.cc:574]: Done

File: dali/test/python/test_pipeline.py (hunk 23 above)
Comment (szalpal) [reply dali/test/python/test_pipeline.py:1661]: Fixed (I believe)

File: dali/test/python/test_pipeline.py (hunk 22 above)
Comment (szalpal) [reply dali/test/python/test_pipeline.py:1669]: Done

File: dali/python/nvidia/dali/pipeline.py [hunk 29]
```diff
@@ -218,6 +236,33 @@ def __init__(self, batch_size = -1, num_threads = -1, device_id = -1, seed = -1,
...
+                        f"`output_dtype` must be either: a type from nvidia.dali.types.DALIDataType, a list of these or None. Found type {type(dtype)} in the list.")
+        elif type(output_dtype) is not types.DALIDataType and output_dtype is not None:
+            raise TypeError(
+                f"`output_dtype` must be either: a type from nvidia.dali.types.DALIDataType, a list of these or None. Found type: {type(output_dtype)}.")
+        self._output_dtype = output_dtype
+
+        # Assign and validate output_ndim
+        if type(output_ndim) is list:
+            for ndim in output_ndim:
+                if type(ndim) is not int and ndim is not None:
```
[the remark this answers is not in the document]
Comment (szalpal) [reply dali/python/nvidia/dali/pipeline.py:253]: Done

File: dali/python/nvidia/dali/pipeline.py [hunk 30]
```diff
@@ -218,6 +236,33 @@ def __init__(self, batch_size = -1, num_threads = -1, device_id = -1, seed = -1,
         else:
             raise TypeError("Expected prefetch_queue_depth to be either int or Dict[int, int]")
 
+        # Assign and validate output_dtype
+        if type(output_dtype) is list:
```
[the remark this answers is not in the document]
Comment (szalpal) [reply dali/python/nvidia/dali/pipeline.py:240]: Done

File: dali/test/python/test_pipeline.py (hunk 20 above)
Comment (szalpal) [reply dali/test/python/test_pipeline.py:1719]: Done

File: dali/python/nvidia/dali/pipeline.py (hunk 18 above)
Comment (szalpal) [reply dali/python/nvidia/dali/pipeline.py:244]: Done

File: dali/pipeline/pipeline.cc [hunk 31]
```diff
@@ -533,8 +541,14 @@ void Pipeline::Build(vector<std::pair<string, string>> output_names) {
   built_ = true;
 }
 
-void Pipeline::SetOutputNames(const vector<std::pair<string, string>> &output_names) {
-  output_names_ = output_names;
+
+void Pipeline::SetOutputDescs(const vector<std::pair<string, string>> &output_names) {
+  DALI_ENFORCE(output_descs_.empty(), "Resetting output_descs_ is forbidden.");
```
[the remark this answers is not in the document]
Comment (szalpal) [reply dali/pipeline/pipeline.cc:546]: Done

File: dali/pipeline/pipeline.cc [hunk 32]
```diff
@@ -533,8 +541,14 @@ void Pipeline::Build(vector<std::pair<string, string>> output_names) {
...
+  output_descs_ = {output_names.begin(), output_names.end()};
+}
+
+void Pipeline::SetOutputDescs(std::vector<PipelineOutputDesc> output_descs) {
```
[the remark this answers is not in the document]
Comment (szalpal) [reply dali/pipeline/pipeline.cc:550]: Well, these are two distict types after all. I don't believe it would be a problem

File: dali/python/nvidia/dali/pipeline.py [hunk 33]
```diff
@@ -218,6 +236,37 @@ def __init__(self, batch_size = -1, num_threads = -1, device_id = -1, seed = -1,
         else:
             raise TypeError("Expected prefetch_queue_depth to be either int or Dict[int, int]")
 
+        # Assign and validate output_dtype
+        if isinstance(output_dtype, (list, tuple)):
+            for dtype in output_dtype:
+                if not(isinstance(dtype, (types.DALIDataType, type(None)))):
+                    raise TypeError(
+                        f"`output_dtype` must be either: a value from nvidia.dali.types.DALIDataType, a list of these or None. Found type {type(dtype)} in the list.")
+                if dtype is not None and dtype == types.NO_TYPE:
```
Comment (stiepan) [inline dali/python/nvidia/dali/pipeline.py:245]:
```suggestion
                if dtype == types.NO_TYPE:
```

File: dali/python/nvidia/dali/pipeline.py [hunk 34]
```diff
@@ -218,6 +236,37 @@ def __init__(self, batch_size = -1, num_threads = -1, device_id = -1, seed = -1,
         else:
             raise TypeError("Expected prefetch_queue_depth to be either int or Dict[int, int]")
 
+        # Assign and validate output_dtype
+        if isinstance(output_dtype, (list, tuple)):
+            for dtype in output_dtype:
+                if not(isinstance(dtype, (types.DALIDataType, type(None)))):
```
Comment (stiepan) [inline dali/python/nvidia/dali/pipeline.py:242]:
```suggestion
                if not isinstance(dtype, (types.DALIDataType, type(None))):
```

Comment (stiepan) [reply dali/python/nvidia/dali/pipeline.py:242]: (similarly below)

File: dali/python/nvidia/dali/pipeline.py [hunk 35]
```diff
@@ -218,6 +236,37 @@ def __init__(self, batch_size = -1, num_threads = -1, device_id = -1, seed = -1,
...
+                    raise TypeError(
+                        f"`output_dtype` must be either: a value from nvidia.dali.types.DALIDataType, a list of these or None. Found type {type(dtype)} in the list.")
+                if dtype is not None and dtype == types.NO_TYPE:
+                    raise ValueError(f"`output_dtype` can't be a types.NO_TYPE. Found {dtype} in the list.")
+        elif not(isinstance(output_dtype, (types.DALIDataType, type(None)))):
+            raise TypeError(
+                f"`output_dtype` must be either: a value from nvidia.dali.types.DALIDataType, a list of these or None. Found type: {type(output_dtype)}.")
+        elif output_dtype is not None and output_dtype == types.NO_TYPE:
```
Comment (stiepan) [inline dali/python/nvidia/dali/pipeline.py:250]:
```suggestion
        elif output_dtype == types.NO_TYPE:
```

File: dali/pipeline/pipeline.cc [hunk 36]
```diff
@@ -708,21 +739,18 @@ string Pipeline::SerializeToProtobuf() const {
...
+    out->set_name(output.name);
+    out->set_device(output.device);
     out->set_is_argument_input(false);
+    out->set_dtype(output.dtype);
+    out->set_ndim(output.ndim);
   }
   pipe.set_device_id(this->device_id_);
   string output = pipe.SerializeAsString();
 
-#ifndef NDEBUG
```
Comment (stiepan) [inline dali/pipeline/pipeline.cc:721]: It's just unrelated clean-up, is it?

**Review by stiepan — APPROVED** [review]

File: dali/test/python/test_pipeline.py [hunk 37]
```diff
@@ -1654,3 +1654,132 @@ def test_invoke_serialize_error_handling_string():
...
+
+def check_dtype_ndim(dali_pipeline, output_dtype, output_ndim, n_outputs):
+    def ndim_dtype_matches(test_value, ref_value):
+        ref_value = ref_value if isinstance(ref_value, (list, tuple)) else [ref_value] * n_outputs
+        if len(ref_value) != len(test_value):
+            return False
+        for tv, ref in zip(test_value, ref_value):
+            if tv != ref:
+                return False
+        return True
```
Comment (stiepan) [inline dali/test/python/test_pipeline.py:1667]: Just crossed my mind, maybe not the most efficient but more "declarative". 

```suggestion
        return list(ref_value) == list(test_value)
```

File: dali/pipeline/pipeline.cc (hunk 31 above)
[the remark this answers is not in the document]
Comment (klecki) [reply dali/pipeline/pipeline.cc:546]: Not done.

File: dali/pipeline/pipeline.cc (hunk 36 above)
Comment (szalpal) [reply dali/pipeline/pipeline.cc:721]: Yes :)

File: dali/test/python/test_pipeline.py (hunk 37 above)
Comment (szalpal) [reply dali/test/python/test_pipeline.py:1667]: Yep, that's prolly way better ;)

**Comment by szalpal** [comment]
!build

**Comment by szalpal** [comment]
!build

**Comment by szalpal** [comment]
!build

**Review by klecki — APPROVED** [review]

**Comment by szalpal** [comment]
!build

**Comment by szalpal** [comment]
!build

**Comment by szalpal** [comment]
!build
